src/Elastic.Apm/Helpers/WildcardMatcher.cs

```diff
@@ -0,0 +1,284 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Elastic.Apm.Helpers
+{
+	/// <summary>
+	/// Heavily inspired by the Java Elastic APM Agent.
+	/// Wildcard matcher to e.g. sanitize strings.
+	/// Examples: `/foo/*/bar/*/baz*`, `*foo*`.
+	/// Matching is case insensitive by default.
+	/// Prepending an element with `(?-i)` makes the matching case sensitive.
+	/// </summary>
+	internal abstract class WildcardMatcher
+	{
+		private const string CaseInsensitivePrefix = "(?i)";
+
+		private const string CaseSensitivePrefix = "(?-i)";
+
+		private const string Wildcard = "*";
+
+		public abstract string GetMatcher();
+
+		/// <summary>
+		/// Checks if the given string matches the wildcard pattern.
+		/// </summary>
+		/// <param name="s">The string to match</param>
+		/// <returns>Whether the string matches the given pattern</returns>
+		public abstract bool Matches(string s);
+
+		/// <summary>
+		/// This is a different version of <see cref="Matches(string)"/> which has the same semantics as calling
+		/// <see cref="Matches(string)" param=" firstPart + secondPart" />.
+		/// The difference is that this method does not allocate memory.
+		/// </summary>
+		/// <param name="firstPart">The first part of the string to match against.</param>
+		/// <param name="secondPart">The second part of the string to match against.</param>
+		/// <returns><code>true</code> when the wildcard pattern matches the partitioned string, <code>false</code> otherwise. </returns>
+		internal abstract bool Matches(string firstPart, string secondPart);
+
+		/// <summary>
+		/// Constructs a new <see cref="WildcardMatcher"/> via a wildcard string.
+		/// By default, matches are a case insensitive.
+		/// It supports the <code>*</code> wildcard which matches zero or more characters.
+		/// Prepend <code>(?-i)</code> to your pattern to make it case sensitive.
+		///
+		/// Example: <code>(?-i)foo*</code> matches the string <code>foobar</code> but does not match <code>FOOBAR</code>.
+		/// </summary>
+		/// <param name="wildcardString"></param>
+		/// <returns></returns>
+		public static WildcardMatcher ValueOf(string wildcardString)
```

Comment:
I think a short circuit could be introduced for exact match cases that can defer to builtin string compare operations (case sensitive or not). e.g a third matcher `VerbatimMatcher`.

Reply:
You mean something like [this](https://github.com/elastic/apm-agent-dotnet/pull/549/commits/20140b503584c3138ea0e210f7e2675baee2ef2b#diff-3d54c6ce902fd9ab1b0ccdd12bb278a9R219)? 

I ran some benchmarks (that is also in the PR) - 1000 long string, `MatcherVerbatim*` use the new `VerbatimWildcardMatcher`, `MatcherSimple*` use the `SimpleWildcardMatcher`:

```
|                         Method |            Mean |          Error |          StdDev |          Median | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------------------- |----------------:|---------------:|----------------:|----------------:|------------:|------------:|------------:|--------------------:|
|   MatcherVerbatimCaseSensitive |        13.36 ns |      0.1536 ns |       0.1282 ns |        13.34 ns |      0.0051 |           - |           - |                32 B |
| MatcherVerbatimCaseInsensitive |        13.30 ns |      0.1715 ns |       0.1605 ns |        13.29 ns |      0.0051 |           - |           - |                32 B |
|     MatcherSimpleCaseSensitive | 2,992,389.11 ns | 59,774.9240 ns | 103,108.8070 ns | 2,941,651.39 ns |      7.8125 |           - |           - |             64032 B |
|   MatcherSimpleCaseInsensitive | 2,896,890.25 ns | 33,103.8968 ns |  29,345.7422 ns | 2,891,317.69 ns |      7.8125 |           - |           - |             64032 B |
```

Based on the numbers: yeah, this probably makes some sense 🤣 .

Reply:
Sweet! I wonder if we can do better on the Simple/Composite matcher paths too. I am happy for that to happen on a follow up PR though